### ISSUE TITLE
Nullptr when costing memory used in memtable to block cache 

### ISSUE THREAD

**Issue by Myasuka** [facebook/rocksdb#6247]
### Expected behavior
Share a write buffer manager among RocksDB instances, and cost memory used in memtable to block cache via [guide](https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager#cost-memory-used-in-memtable-to-block-cache).
I use RocksJava (RocksDB-5.18.3) to achieve this and because I want to limit the memory usage strictly, the cache I created is strictly limited like:
~~~ java
final Cache cache = new LRUCache(size, -1, true, highPriorityPoolRatio);
final WriteBufferManager wbm = new WriteBufferManager((long) (writeBufferRatio * size), cache);
~~~

### Actual behavior
However, the java process would crash with core dump when I put entries in RocksDB. Below is the stack trace:
~~~ shell
C  [librocksdbjni-linux64.so+0x200fa0]  rocksdb::LRUCache::GetHash(rocksdb::Cache::Handle*) const+
C  [librocksdbjni-linux64.so+0x3c0379]  rocksdb::WriteBufferManager::FreeMemWithCache(unsigned long)+0xa9
C  [librocksdbjni-linux64.so+0x3b8795]  rocksdb::AllocTracker::FreeMem()+0x65
C  [librocksdbjni-linux64.so+0x30a973]  rocksdb::MemTable::~MemTable()+0x23
C  [librocksdbjni-linux64.so+0x21be9e]  rocksdb::SuperVersion::~SuperVersion()+0x5e
C  [librocksdbjni-linux64.so+0x26f2b3]  rocksdb::SuperVersionContext::Clean()+0x1b3
C  [librocksdbjni-linux64.so+0x26f3af]  rocksdb::JobContext::Clean()+0x6f
C  [librocksdbjni-linux64.so+0x28b029]  rocksdb::DBImpl::BackgroundCallFlush(rocksdb::Env::Priority)+0x1e9
C  [librocksdbjni-linux64.so+0x494cf4]  rocksdb::ThreadPoolImpl::Impl::BGThread(unsigned long)+0x254
C  [librocksdbjni-linux64.so+0x494e8d]  rocksdb::ThreadPoolImpl::Impl::BGThreadWrapper(void*)+0x5d

~~~ 

### Steps to reproduce the behavior
I try to analysis why this happens.
Current write buffer manager implementation with dummy entries is [below](https://github.com/facebook/rocksdb/blob/1369736b18598f9dd4c1d33df017926b8e1514b1/memtable/write_buffer_manager.cc#L92):
~~~ cpp
  while (new_mem_used > cache_rep_->cache_allocated_size_) {
    // Expand size by at least 1MB.
    // Add a dummy record to the cache
    Cache::Handle* handle;
    cache_rep_->cache_->Insert(cache_rep_->GetNextCacheKey(), nullptr,
                               kSizeDummyEntry, nullptr, &handle);
    cache_rep_->dummy_handles_.push_back(handle);
    cache_rep_->cache_allocated_size_ += kSizeDummyEntry;
  }
~~~
As you can see, the returned status of Insert is ignored, and we would push back a `nullptr` handle into `dummy_handles` when cache exceeds the capacity. That's why when we free memory, and call `GetHash` to the `nullptr` leading to the crash.

Moreover, as `WriteBufferManager::ReserveMemWithCache` is a internal action, how can we throw the exception out to java environment?

**Comment by Myasuka** [comment]
@siying @adamretter could you take a look at this problem?

**Comment by Myasuka** [comment]
@siying , I think this problem also existed in pure-c++ RocksDB when we enable the strict limit of block cache. Do you have any thoughts about this problem?

**Comment by siying** [comment]
@Myasuka is the problem fixed?

**Comment by Myasuka** [comment]
@siying thanks for your fix and this problem has gone with latest RocksDB-6.11.4

**Comment by Myasuka** [comment]
@siying when I try to combine strict limit block cache with writ buffer manager, I met another issue (reported in https://github.com/facebook/rocksdb/issues/8670).
Do you think memory control could get better with strict limit block cache if resolved the issue #8670?
